### ISSUE TITLE
[question] How to make rootfs libraries from toolchain package transparent

### ISSUE THREAD

**Issue by nicosmd** [conan-io/conan#12597]
We are developing an application which is compiled for x86 and cross compiled for arm. The app has a lot of dependencies like for example `OpenCV`, which we now want to manage using conan. For the arm build we've already created a toolchain package containing the cross compiler toolchain and some libraries in a rootfs directory, which are produced during the build of our linux image of the embedded device. These libraries are coming from our SoC's vendor and are also depending for example on `OpenCV`, why its libraries are part of the rootfs as well. For the x86 build we want to completely rely on the `OpenCV` package which are available at conan center. For the application we are already using `CMakeToolchain` and `CMakeDeps` generator.

The problem we currently have is that in the applications `conanfile.py` we basically want to have 

```
    def requirements(self):
         self.requires("opencv/4.5.5")
```

which is absolutely fine for the x86 build because there we don't rely on these vendor libraries. But in context of arm I think it is a bad idea to link both the conan center `OpenCV` and the toolchains `OpenCV` (coming transitively from the vendors libraries) to our application. The only solution, which I can see here, is to only rely on the `OpenCV` library from the toolchain package.

The idea I have so far was to utilize somehow the `CMakeDeps` generator to normally produce for x86 build the *Config.cmake files inside the CMAKE_BINARY_DIR, which points to the `OpenCV` libraries in the conan cache and for the arm build I create custom *Config.cmake files inside the toolchain package, which points to the `OpenCV` libraries in the toolchain package. As far as I understand, I could use `cmake_build_modules` and `build_context_build_modules` properties to achieve this, so the `find_package` should work in both cases and I do not need to change anything in the CMakeLists.txt.

**My Question now is**: how can I tell conan to **not** generate the \*Config.cmake, which are provided by the toolchain package and how do I make the dependencies provided by the toolchain package transparent to conan?

I think the second part of the question is quite important because libraries like OpenCV have also dependencies to some very common libraries like zlib, which then must be part of the rootfs as well. Imagine there is a package libA, which depends on zlib too, then I want that the build for libA also uses the toolchains zlib and I want to be informed if there is a version conflict somewhere in my dependency graph. I think this can only be achieved when conan knows which libraries with which version are provided by the toolchain package, otherwise all the dependencies of the firmware libraries would become marshy and I'm again in the dependency hell. 

Maybe I'm over-complicating things, but I have the feeling that this must be a common use case to have some libraries be part of the rootfs in embedded systems and that there is already an obvious solution I'm missing.

Thank you very much in advance!

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @nicosmd 

This is a very interesting question.
I am trying to understand the setup, have you considered something like this?

```python
def requirements(self):
      if self.settings.arch == "x86":
              self.requires("opencv/....")

def build_requirements(self):
      if self.settings.arch == "armv8":
              self.tool_requires("myarmtoolchain/....")

def generate(self):
      if self.settings.arch == "x86":
              cmakedeps = CMakeDeps(self)
              cmakedeps.generate()
```

I mean, it seems that you have 2 scenarios, in the x86 one, you want dependencies to be managed by Conan, but in the ``armv8`` the dependencies will *not* come from Conan, but bundled in the rootfs. So expressing this conditionality on your recipes could be the way. (it is difficult to know, just an idea to try to understand the problem better)

**Comment by nicosmd** [comment]
Hey @memsharded,

Thank you very much for your quick response!

Let me add some details to the context. 

When building for arm, the app _not only_ rely on the toolchain rootfs, so there are some OSS Dependencies which are not part of the rootfs but which are consumed by our app. Beside that we also have to retrieve several internal component packages which are developed by other teams in our project. 

To add this to your suggestion it could look like this, I guess:

```python
def requirements(self):
      self.requires("componentA/...") # developed in our project
      self.requires("protobuf/...") # OSS Dependency not in rootfs
      if self.settings.arch == "x86":
              self.requires("opencv/....")

def build_requirements(self):
      if self.settings.arch == "armv8":
              self.tool_requires("myarmtoolchain/....") # we are setting this in our profile

def generate(self):
      cmakedeps = CMakeDeps(self)
      cmakedeps.generate()
```
So for arm we will have a mixed scenario. Some libraries come from conan, some other from the toolchain package.

The problem then is that protobuf and opencv are relying both on zlib, wheras opencv will link the zlib library from the toolchain package, which is most likely a shared library, and  protobuf will try to retrieve it as a conan package, which could be a static zlib library with a different version. This might never create some issues in this particular case, nevertheless this feels a bit unsafe. We also do not want to put everything inside the rootfs, because we want to use the advantage of conan to bring in dependencies easily without the need to create a new linux image all the time. So we only want to use the toolchain libraries, which must be available in the rootfs due to these vendor libraries. 

Another issue which come to my mind is that we need to add this dependency toggle mechanism to every package where packages are required, which are available in the toolchain as well. In the future we will have very likely more platforms for which we need to deliver our app and the rootfs there could again look different, which feels a bit like manual dependency management in the end, which we would like to avoid but could life with.

My vision would be that the conanfile of our app would contain 

```python
def requirements(self):
      self.requires("componentA/...") # developed in our project
      self.requires("protobuf/...") # OSS Dependency not in rootfs
      self.requires("opencv/....")

def build_requirements(self):
      if self.settings.arch == "armv8":
              self.tool_requires("myarmtoolchain/....") # we are setting this in our profile

def generate(self):
      cmakedeps = CMakeDeps(self)
      cmakedeps.generate()
```

and we can somehow tell conan, particularly the CMakeDeps generator, that for the arm build it does not need to generate the .cmake files for opencv, because it is provided by our toolchain package. And in the toolchain package I would like to specify which opencv version it provides so that the dependecy graph will not contain any gaps so it is also transparent for protobuf that there is maybe another zlib library in the graph with a different version.

I hope this helps to understand our setup a bit better and what we would like to achieve somehow. Thanks again!

**Comment by memsharded** [comment]
This seems this need could be addressed by https://github.com/conan-io/conan/issues/7118, in which we will propose a mechanism to tell ``CMakeDeps`` directly in the consumer side, what files should be generated for which dependencies.

Still the ``opencv`` requirement if it is going to be in the ``rootfs`` shouldn't be declared as ``requires`` for that architecture, unless there are 2 opencvs, one you manage as a conan package that you want to ``find_package()`` and the other in the ``rootfs`` that is there just for the toolchain to work, but transparent to your usage.

**Comment by nicosmd** [comment]
Hey @memsharded , sorry for the late response. It took me some time to get into the idea but now the feature seems to be integrated in 1.55. 

I guess you are right that this feature could help in our scenario. But what I did not understand: when I need to exclude `opencv` from the requirements for this architecture anyway, do I even need to change the properties for it using the new feature? Probably, I did not fully understand how to utilize it in our scenario. 

Maybe I have some general misconception but wouldn't it be a problem to simultaneously link a library two times into our application? I understand that I can just exclude `opencv` as a requirement for that architecture and find_package will find the dependency's *Config.cmake  in the toolchain package. But what is about the `zlib` dependency coming from `protobuf`? We would like to use the OSS conan packages as they are without changing them. Currently, I can't see any way to avoid to add

```python
      if self.settings.arch == "x86":
              self.requires("zlib/....")
``` 

to the conanfile.py of `protobuf` in order to rebuild `protobuf` with the `zlib` from the `rootfs` for the arm architecture. So we would need to maintain our own conanfiles for every of these "critical" dependencies. Otherwise we would probably link `zlib` twice, once from the toolchain transitively linked by `opencv` and once from the conan cache transitivley linked by `protobuf`, which might be not good. Am I right or is there a smart way to avoid changing the conanfile? Maybe by using the new feature?

**Comment by nicosmd** [comment]
Just a follow up on this from my side, since we might found a solution, where we could need some feedback. But there also seems to be two ongoing somehow related feature requests, which I wanted to mention here as well.

First of all, the solution we have found for us: For every oss library, which is part of our dependency graph and is available as a conan package in conan center as well as in the vendor sysroot, we have created a new conanfile. This conanfile only packages the prebuild library from that sysroot and we set the version to something generic. If we use zlib as an example, we name it for example: zlib/sysroot. So there are basically two conanfiles, which exists in parallel satisfying the same dependency. This zlib/sysroot package is now set as a requirement in the sysroot conan package. In this way we achieve to force any consumer of this library to override the zlib dependency, when there is another package in the graph using zlib. Our product application conan packages looks then basically like this: 

```python
def requirements (self):
    self.requires("opencv/4.5.5")
    ....
    if self.settings.arch == "armv8":
        self.requires("vendor/sysroot")
        self.requires("zlib/sysroot", override = True)
``` 
In this way we are able to model what we need. 

But I'm still not completely satisfied by this solution. First, we need to add this additional conanfile, where i'm not sure, if this really is a good idea or maybe a bad practise. Second, we can't prebuild for example opencv in isolation for the armv8 platform and always need to recompile it, when we are building our product package. 

I came across two feature request, which might go in the same direction: #12459 and #12478 . Maybe this cmake dependency provider is a good solution to model our need. But this will only work, when every conflicting oss dependency is using cmake itself, which might not be the case. I was asking myself if both issues couldn't be satisfied by some feature, which would allow something like in a sysroot conan package :

```python
def requirements(self):
   self.provides("zlib/1.12.1")
```
in order to model, that a toolchain or sysroot package provides an actual dependency, so that the conan generators could use this as as additional information. 

I'm very curious what you think about our solution and this "provider" concept. Thank's in advance!
